#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [87]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import median_absolute_error
from sklearn.metrics import explained_variance_score
from sklearn.metrics import max_error
import urllib.request
from PIL import Image

## Data

In [92]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/train 1.csv", index_col = 0)
test = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 1.csv", index_col = 0)


## Data exploration

In [89]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 912 entries, 0 to 911
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Ram               912 non-null    int64  
 1   Weight            912 non-null    float64
 2   Memory            912 non-null    int64  
 3   ScreenResolution  912 non-null    int64  
 4   Price_euros       912 non-null    float64
dtypes: float64(2), int64(3)
memory usage: 42.8 KB


In [93]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 391 entries, 0 to 390
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Ram               391 non-null    int64  
 1   Weight            391 non-null    float64
 2   Memory            391 non-null    int64  
 3   ScreenResolution  391 non-null    int64  
dtypes: float64(1), int64(3)
memory usage: 15.3 KB


In [90]:
train.head()

Ram  Weight  Memory  ScreenResolution  Price_euros
0    4    2.40     500           1049088        387.0
1    8    1.23     256           5760000       1379.0
2    8    3.00     256           2073600        854.0
3    8    1.26     512           2073600       1483.0
4    8    2.80    1024           1440000        519.0

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [99]:
X = train.drop("Price_euros", axis = 1)
y = train['Price_euros']
X.shape

(912, 4)

In [100]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [42]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

### 3. Asignar el modelo (vacío) a una variable


In [101]:
lm = LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [102]:
lm.fit(X, y)

LinearRegression()

In [103]:
coeff_df = pd.DataFrame(lm.coef_,
                        X.columns,
                        columns=['Coefficient'])
coeff_df.sort_values('Coefficient', ascending=False)

Coefficient
Ram                 95.814941
ScreenResolution     0.000116
Memory              -0.172698
Weight              -1.479176

In [104]:
predictions = lm.predict(X)
predictions

array([ 595.75947746, 1568.69751318, 1138.97066251, 1097.33375365,
        933.22516434, 1773.59775124, 1140.89359095,  949.44121086,
       1568.69751318,  983.71558985, 1729.38707662,  661.14207932,
       1568.69751318, 1141.48526124,  624.26221688, 1141.48526124,
       1054.33753249,  595.75947746, 1286.19653752, 1140.31671242,
       2583.4158853 , 1726.54705924, 1006.85635016, 1007.34447815,
       1141.76630463,  505.41367042, 1283.86839546, 1141.3965107 ,
       1140.99713325,  779.59124833, 1141.500053  , 1140.9823415 ,
        596.0553126 , 1908.03436947, 1751.49241393, 1906.71790308,
       1140.59775581,  714.94823433,  596.0553126 , 1728.64748876,
        596.0553126 , 1862.83264712, 3215.37543406,  677.64712351,
        791.53142242, 1140.15400309,  550.6162393 , 1051.33285077,
        506.06450774,  672.16516446, 1524.4720468 ,  661.6302073 ,
       1818.72551481,  596.0553126 , 1141.69234584, 2202.48147174,
        716.14636667, 1141.36692718, 1141.33734367, 1140.96754

In [105]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error ,mean_squared_error, r2_score

print("MAE: ", mean_absolute_error(y, predictions))
print("MAPE: ", mean_absolute_percentage_error(y,predictions)) # Es el complemento del R2
print("MSE: ", mean_squared_error(y,predictions))
print("RMSE: ", np.sqrt(mean_squared_error(y,predictions))) #es el mas/menos en la misma unidad que el target
print("R2: ", r2_score(y,predictions))

MAE:  312.2654825960655
MAPE:  0.3557905650822285
MSE:  178144.290442125
RMSE:  422.07142812813686
R2:  0.6161473241825448


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

In [49]:
# from sklearn.preprocessing import PolynomialFeatures
# poly_feats = PolynomialFeatures(degree = 3)
# poly_feats.fit(X)
# X_poly = poly_feats.transform(X)
# X_poly




# pol_reg = LinearRegression()
# pol_reg.fit(X_train, y_train)


# predictions = pol_reg.predict(X_test)


# df_preds = pd.DataFrame(predictions, columns = ['predictions'])
# df_preds['real_value'] = pd.Series(y_test)

# df_preds

# from sklearn.metrics import r2_score
# r2_score(pol_reg.predict(X_test), y_test)

### 2. Carga los datos de `test.csv` para predecir.

In [106]:
# X_pred = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 1.csv", index_col = 0)
X_pred = test
X_pred.head()

Ram  Weight  Memory  ScreenResolution
0    8    1.25     256           2073600
1    4    1.40      32           2304000
2    2    1.40      32           2073600
3    8    2.65    1024           2073600
4    2    1.10      32           1049088

In [107]:
X_pred.shape

(391, 4)

**¿Por qué da error?** 

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [111]:
X_pred.head()

Ram  Weight  Memory  ScreenResolution
0    8    1.25     256           2073600
1    4    1.40      32           2304000
2    2    1.40      32           2073600
3    8    2.65    1024           2073600
4    2    1.10      32           1049088

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [112]:
predictions_submit = lm.predict(X_pred)
predictions_submit

array([1141.55922003,  823.45621595,  605.13204042, 1006.85635016,
        486.87516425,  756.89424073,  624.55805202, 1750.44219917,
       1862.16701805,  984.97288921, 1140.65692284, 1140.99713325,
        985.26872436, 1021.4533742 ,  779.428539  , 1140.68650635,
       1141.61838706, 1907.04332174,  624.26221688, 1007.52197924,
       1154.01863498,  984.69184583,  595.92218679, 1021.88233516,
        660.8018689 , 1284.32693993,  697.20626093,  833.93652608,
        602.15687957,  984.97288921,  985.68289356,  963.60713977,
        948.22828677,  963.01546948, 1183.27892927, 1141.85505517,
        985.71247707, 1141.04150852,  716.50798832, 1863.7053608 ,
       1726.72456032,  984.97288921,  680.75050109, 1375.6983027 ,
       1140.30192066, 1728.73623931, 1007.15218531, 1728.64748876,
       1139.51795753, 1861.61972303,  714.59323216,  624.18825809,
       1007.46281221, 1140.62733932,  660.29894915,  757.13090885,
        985.12080679,  678.28316907, 1141.42609421, 1141.18942

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [114]:
sample = pd.read_csv("data/sample_submission.csv")

In [115]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [116]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [117]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [118]:
submission.head()

laptop_ID  Price_euros
0        750  1141.559220
1        726   823.456216
2        633   605.132040
3        363  1006.856350
4        319   486.875164

In [119]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [120]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [121]:
chequeator(submission)

You're ready to submit!
